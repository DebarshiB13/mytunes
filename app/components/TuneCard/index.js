/**
 *
 * TuneCard
 *
 */

import React, { memo, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Card, Typography, Image, Button } from 'antd';
import { PlayCircleFilled, PauseCircleFilled } from '@ant-design/icons';
import If from '@components/If';
import { Link } from 'react-router-dom';

const { Title } = Typography;

const ItemCard = styled(Card)`
  && {
    margin: 20px 0;
    display: flex;
    width: ${(props) => props.maxwidth}px;
    flex-direction: column;
    justify-content: space-around;
  }
`;

const iconStyle = { fontSize: 32 };

const IconButton = styled(Button)`
  display: flex;
  justify-content: center;
  align-items: center;
  position: relative;
`;

const StyledT = styled(Title)`
  && {
    font-size: ${(props) => props.fontSize}em;
  }
`;

export function TuneCard({ maxwidth, artistName, collectionName, cardImg, previewUrl, handleOnActionClick, songId }) {
  const audioRef = useRef();
  const [play, setPlay] = useState(false);

  const handlePlayPause = (e) => {
    e.preventDefault();
    const isPaused = audioRef.current.paused;
    if (isPaused) {
      audioRef.current.play();
    } else {
      audioRef.current.pause();
    }
    setPlay(!play);
    if (handleOnActionClick) {
      handleOnActionClick(audioRef);
    }
  };

  const renderItemCard = () => (
    <ItemCard maxwidth={maxwidth} data-testid="tune-card">
      <StyledT fontSize={1.6} data-testid="artist-name">
        {artistName}
      </StyledT>
      <Image src={cardImg} width="100%" preview={false} />
      <StyledT fontSize={1.2}>{collectionName}</StyledT>

      <IconButton
        shape="circle"
        type="text"
        data-testid="play-pause-btn"
        onClick={handlePlayPause}
        icon={
          <If
            condition={!audioRef.current?.paused && audioRef.current?.src}
            otherwise={<PlayCircleFilled style={iconStyle} />}
          >
            <PauseCircleFilled style={iconStyle} />
          </If>
        }
      />
      <audio ref={audioRef} data-testid="audio" src={previewUrl}></audio>
    </ItemCard>
  );

  return (
    <If condition={songId} otherwise={renderItemCard()}>
      <Link to={`/songs/${songId}`} songId={songId}>
        {renderItemCard()}
      </Link>
    </If>
  );
}

TuneCard.defaultProps = {
  maxwidth: 300
};
TuneCard.propTypes = {
  artistName: PropTypes.string,
  maxwidth: PropTypes.number,
  collectionName: PropTypes.string,
  cardImg: PropTypes.string,
  previewUrl: PropTypes.string,
  handleOnActionClick: PropTypes.func,
  songId: PropTypes.number
};

export default memo(TuneCard);
