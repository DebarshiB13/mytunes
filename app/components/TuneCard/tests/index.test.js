/**
 *
 * Tests for TuneCard
 *
 */

import React from 'react';
import { fireEvent } from '@testing-library/dom';
import { renderWithIntl } from '@utils/testUtils';
import TuneCard from '../index';

describe('<TuneCard />', () => {
  it('should render and match the snapshot', () => {
    const artistName = 'Alan Walker';
    const collectionName = 'Some Collection';
    const { baseElement } = renderWithIntl(<TuneCard artistName={artistName} collectionName={collectionName} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should call handlePlay on Click', () => {
    const { getByTestId } = renderWithIntl(<TuneCard />);
    const handlePlay = jest.fn();
    fireEvent.click(getByTestId('play-pause-btn'), { onclick: handlePlay() });
    expect(handlePlay).toHaveBeenCalled();
  });

  it('should play/pause audio when button is clicked', () => {
    const playStub = jest.spyOn(window.HTMLMediaElement.prototype, 'play').mockImplementation(() => {});
    const pauseStub = jest.spyOn(window.HTMLMediaElement.prototype, 'pause').mockImplementation(() => {});

    const { getByTestId } = renderWithIntl(<TuneCard />);

    const button = getByTestId('play-pause-btn');

    fireEvent.click(button);
    expect(playStub).toHaveBeenCalled();

    fireEvent.click(button);
    expect(pauseStub).toHaveBeenCalled();

    playStub.mockRestore();
    pauseStub.mockRestore();
  });

  it('should set/unset audio url on play/pause button click', () => {
    const songPreviewlUrl =
      'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview115/v4/f9/da/66/f9da6605-fafe-6e21-6e11-e58f2221b7bf/mzaf_5280154181764903840.plus.aac.p.m4a';

    const { getByTestId } = renderWithIntl(<TuneCard previewUrl={songPreviewlUrl} />);
    let audio = getByTestId('audio');
    let button = getByTestId('play-pause-btn');

    expect(audio.src).toEqual('');
    fireEvent.click(button);
    expect(audio.src).toEqual(songPreviewlUrl);
    fireEvent.click(button);
    expect(audio.src).not.toEqual(songPreviewlUrl);
  });
});
