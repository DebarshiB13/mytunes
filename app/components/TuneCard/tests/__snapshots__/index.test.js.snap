// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TuneCard /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="ant-card ant-card-bordered TuneCard__ItemCard-sc-1cl6pnu-0 dBepYt"
      data-testid="tune-card"
      maxwidth="300"
    >
      <div
        class="ant-card-body"
      >
        <h1
          class="ant-typography TuneCard__StyledT-sc-1cl6pnu-2 eJcBSi"
          data-testid="artist-name"
          font-size="1.6"
        >
          AlanWalker
        </h1>
        <div
          class="ant-image"
          style="width: 100%;"
        >
          <img
            class="ant-image-img"
            src="https://is4-ssl.mzstatic.com/image/thumb/Music115/v4/8d/af/2e/8daf2ed5-b5c1-9c53-17fc-a10d7a247121/source/100x100bb.jpg"
          />
        </div>
        <h1
          class="ant-typography TuneCard__StyledT-sc-1cl6pnu-2 hVdOjy"
          font-size="1.2"
        >
          Some Collection
        </h1>
        <button
          class="ant-btn ant-btn-text ant-btn-circle ant-btn-icon-only TuneCard__IconButton-sc-1cl6pnu-1 eahDeJ"
          data-testid="play-pause-btn"
          type="button"
        >
          <span
            aria-label="play-circle"
            class="anticon anticon-play-circle"
            role="img"
            style="font-size: 32px;"
          >
            <svg
              aria-hidden="true"
              data-icon="play-circle"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm144.1 454.9L437.7 677.8a8.02 8.02 0 01-12.7-6.5V353.7a8 8 0 0112.7-6.5L656.1 506a7.9 7.9 0 010 12.9z"
              />
            </svg>
          </span>
        </button>
        <audio
          data-testid="audio"
          src="https://abc.mp3/"
        />
      </div>
    </div>
  </div>
</body>
`;
